pkg/controller/atlasproject/maintenancewindow.go

```diff
@@ -0,0 +1,164 @@
+// in Atlas but is not specified in the CR, it is deleted.
+func ensureMaintenanceWindow(ctx *workflow.Context, projectID string, atlasProject *mdbv1.AtlasProject) workflow.Result {
+	windowSpec := atlasProject.Spec.ProjectMaintenanceWindow
+	if err := validateMaintenanceWindow(windowSpec); err != nil {
+		return workflow.Terminate(workflow.ProjectWindowInvalid, err.Error())
+	}
+
+	if isEmptyWindow(windowSpec) {
+		ctx.Log.Debugw("Window empty or undefined, deleting in Atlas")
+		if result := deleteInAtlas(ctx.Client, projectID); !result.IsOk() {
+			return result
+		}
+		return workflow.OK()
+	}
+
+	if windowSpecified(windowSpec) {
+		ctx.Log.Debugw("Checking if window needs update")
+		windowInAtlas, result := getInAtlas(ctx.Client, projectID)
+		if !result.IsOk() {
+			return result
+		}
+
+		if windowInAtlas.DayOfWeek == 0 || windowInAtlas.HourOfDay == nil ||
+			*windowInAtlas.HourOfDay != windowSpec.HourOfDay || windowInAtlas.DayOfWeek != windowSpec.DayOfWeek {
+			ctx.Log.Debugw("Creating or updating window")
+			// We set startASAP to false because the operator takes care of calling the API a second time if both
+			// startASAP and the new maintenance timeslots are defined
+			if result := createOrUpdateInAtlas(ctx.Client, projectID, windowSpec.WithStartASAP(false)); !result.IsOk() {
+				return result
+			}
+		} else if *windowInAtlas.AutoDeferOnceEnabled != windowSpec.AutoDefer {
+			// If autoDefer flag is different in Atlas, and we haven't updated the window previously, we toggle the flag
+			ctx.Log.Debugw("Toggling autoDefer")
+			if result := toggleAutoDeferInAtlas(ctx.Client, projectID); !result.IsOk() {
+				return result
+			}
+		}
+	}
+
+	if windowSpec.StartASAP {
+		ctx.Log.Debugw("Starting maintenance ASAP")
+		// To avoid any unexpected behavior, we send a request to the API containing only the StartASAP flag,
+		// although the API should ignore other fields in that case
+		if result := createOrUpdateInAtlas(ctx.Client, projectID, project.NewMaintenanceWindow().WithStartASAP(true)); !result.IsOk() {
+			return result
+		}
+		// Nothing else should be done after sending a StartASAP request
+		return workflow.OK()
+	}
+
+	if windowSpec.Defer {
+		ctx.Log.Debugw("Deferring scheduled maintenance")
+		if result := deferInAtlas(ctx.Client, projectID); !result.IsOk() {
+			return result
+		}
+		// Nothing else should be done after deferring
+		return workflow.OK()
+	}
+
+	return workflow.OK()
+}
+
+func isEmpty(i int) bool {
+	return i == 0
+}
+
+func isEmptyWindow(window project.MaintenanceWindow) bool {
+	return isEmpty(window.DayOfWeek) && isEmpty(window.HourOfDay) && !window.StartASAP && !window.Defer && !window.AutoDefer
+}
+
+func windowSpecified(window project.MaintenanceWindow) bool {
+	return !isEmpty(window.DayOfWeek)
+}
+
+func maxOneFlag(window project.MaintenanceWindow) bool {
+	sum := 0
+	if window.StartASAP {
+		sum++
+	}
+	if window.Defer {
+		sum++
+	}
+	return sum <= 1
+}
```

Comment:
I assume you want to check if only one flag is set? Consider simplifying it like that: `!(window.StartASAP && window.Defer)`

Reply:
Good catch, I've implemented it like that because there were more than 2 flags originally, but now it simplifies the code a lot.